src/shared/serverinfo.hpp

```diff
@@ -0,0 +1,231 @@
+//It have 2 list of members:
+// * row 1: original names, for compat, and maybe to allow some naming support in
+//          CS latter
+// * row 2: aliases to avoid reapeating 'SINFO::' for nothing.
+// TODO: use scoped enums
+enum SINFO
+{
+    SINFO_NONE   ,
+    SINFO_STATUS ,
+    SINFO_NAME   ,
+    SINFO_PORT   ,
+    SINFO_QPORT  ,
+    SINFO_DESC   ,
+    SINFO_MODE   ,
+    SINFO_MUTS   ,
+    SINFO_MAP    ,
+    SINFO_TIME   ,
+    SINFO_NUMPLRS,
+    SINFO_MAXPLRS,
+    SINFO_PING   ,
+    SINFO_PRIO   ,
+    SINFO_MAX    ,
+};
+
+// statuses a server can have, about being joinable
+// TODO: use scoped enums
+enum SSTAT
+{
+    SSTAT_OPEN   ,
+    SSTAT_LOCKED ,
+    SSTAT_PRIVATE,
+    SSTAT_FULL   ,
+    SSTAT_UNKNOWN,
+    SSTAT_MAX    ,
+};
+
+// values returned when trying to join a server?
+// used in m_attr[5]
+enum MM
+{
+    MM_OPEN    ,
+    MM_VETO    ,
+    MM_LOCKED  ,
+    MM_PRIVATE ,
+    MM_PASSWORD,
+};
+
+struct serverinfo
+{
+    // should probably be private, but:
+    // WAITING is used in serverbrowser for cubescript exposition
+    enum
+    {
+        MAXPINGS = 3,
+        WAITING = INT_MAX
+    };
+
+private:
+    // list of player/spectators names
+    std::vector<std::string> m_players;
+    // authentification per player
+    std::vector<std::string> m_handles;
+
+    // various server attributes, see cube_get_property() for what I know
+    std::vector<int> m_attr;
+    // server's UDP/TCP/IP informations
+    ENetAddress m_address;
+    // description as seen in serverbrowser
+    string m_sdesc;
+
+    string m_flags;
+    string m_branch;
+    string m_authhandle;
+
+    // probably the IP or domain name
+    string m_name;
+    // name of current map
+    string m_map;
+
+    int m_lastping;
+    int m_nextping;
+    int m_pings[MAXPINGS];
+    int m_lastinfo;
+
+    // UDP? port number, probably redundant with m_address
+    int m_port;
+
+    int m_priority;
+    int m_ping;
+
+    // name resolution state.
+    enum
+    {
+        UNRESOLVED,
+        RESOLVING,
+        RESOLVED
+    } m_resolved;
+
+    // is "servers" sorted?
+    static bool s_sortedservers;
+
+    const char* description( void ) const;
+    void clearpings();
+    void cleanup();
+    void addping(int rtt, int millis);
+    int protocol_version( void ) const;
+    SSTAT server_status( void ) const;
+public:
+    serverinfo(uint ip, int port, int priority = 0);
+    ~serverinfo();
+
+    // marks server as not queried and flags servers(list) as not sorted
+    void reset();
+
+    // compare 2 servers according to a sorting "style".
+    // returns 0 if style is invalid, abort() in debug mode.
+    //
+    // Styles NUMPLRS and PRIO are "better if higher".
+    // Others are "better if lower".
+    // Reverse invert the sorting.
+    //
+    // returns:
+    //   * -1 if *this < other
+    //   *  0 if *this = other
+    //   *  1 if *this > other
+    int compare( serverinfo const& other, int style, bool reverse ) const;
```

Comment:
Do we absolutely need both overloads? I think one should just cast before use if needed.